llvm/include/llvm/ADT/PagedVector.h

```diff
@@ -0,0 +1,322 @@
+            OldPage == NewPage ? Size % PAGE_SIZE : PAGE_SIZE;
+        for (std::size_t J = NewSize % PAGE_SIZE; J < LastPageElements; ++J)
+          Page[J].~T();
+      }
+      PageToDataIdx.resize(NewLastPage + 1);
+    }
+    Size = NewSize;
+    // If the capacity is enough, just update the size and continue
+    // with the currently allocated pages.
+    if (Size <= capacity())
+      return;
+    // The number of pages to allocate. The Remainder is calculated
+    // for the case in which the NewSize is not a multiple of PAGE_SIZE.
+    // In that case we need one more page.
+    std::size_t Pages = Size / PAGE_SIZE;
+    std::size_t Remainder = Size % PAGE_SIZE;
+    if (Remainder != 0)
+      Pages += 1;
+    assert(Pages > PageToDataIdx.size());
+    // We use InvalidPage to indicate that a page has not been allocated yet.
+    // This cannot be 0, because 0 is a valid page id.
+    // We use InvalidPage instead of a separate bool to avoid wasting space.
+    PageToDataIdx.resize(Pages, InvalidPage);
+  }
+
+  // Return true if the vector is empty
+  [[nodiscard]] bool empty() const { return Size == 0; }
+
+  /// Clear the vector, i.e. clear the allocated pages, the whole page
+  /// lookup index and reset the size.
+  void clear() {
+    Size = 0;
+    // If we own the allocator, simply reset it, otherwise we
+    // deallocate the pages one by one.
+    if (Allocator.getInt() == true)
+      Allocator.getPointer()->Reset();
+    else
+      for (uintptr_t Page : PageToDataIdx)
+        Allocator.getPointer()->Deallocate(reinterpret_cast<T *>(Page));
+
+    PageToDataIdx.clear();
+  }
+
+  // Iterator on all the elements of the vector
+  // which have actually being constructed.
+  class MaterialisedIterator {
+    PagedVector const *PV;
+    std::size_t ElementIdx;
+
+  public:
+    using iterator_category = std::forward_iterator_tag;
+    using value_type = T;
+    using difference_type = std::ptrdiff_t;
+    using pointer = T *;
+    using reference = T &;
+
+    MaterialisedIterator(PagedVector const *PV, std::size_t ElementIdx)
+        : PV(PV), ElementIdx(ElementIdx) {}
+
+    // When incrementing the iterator, we skip the elements which have not
+    // been materialised yet.
+    MaterialisedIterator &operator++() {
+      while (ElementIdx < PV->Size)
+        if (PV->PageToDataIdx[ElementIdx++ / PAGE_SIZE] != InvalidPage)
+          break;
+
+      return *this;
+    }
+    // Post increment operator.
+    MaterialisedIterator operator++(int) {
+      MaterialisedIterator Copy = *this;
+      ++*this;
+      return Copy;
+    }
+
+    std::ptrdiff_t operator-(MaterialisedIterator const &Other) const {
+      assert(PV == Other.PV);
+      // If they are on the same table we can just subtract the indices.
+      // Otherwise we have to iterate over the pages to find the difference.
+      // If a page is invalid, we skip it.
+      if (PV == Other.PV)
+        return ElementIdx - Other.ElementIdx;
```

Comment:
This doesn't seem right to me -- you already asserted the `if` condition here. Should this `if` be checking whether the iterators are on the same page rather than pointing into the same vector?

Reply:
Indeed. I dropped this method in any case, so this does not need fixing anymore.